projects/ti_to_ti/preference_tokenize_example.py

```diff
@@ -0,0 +1,308 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+
+import io
+import os
+import json
+from abc import ABC, abstractmethod
+import copy
+from typing import Any, Callable
+from typing_extensions import TypedDict  # Python 3.10+
+from tqdm import tqdm
+
+import requests
+import librosa
+from PIL import Image
+from torchvision.io import read_video
+import uuid
+
+from align_anything.utils.template_registry import register_template
+
+
+import torch
+import torch.multiprocessing as mp
+
+import transformers
+from torch.utils.data import Dataset
+from torchvision import transforms
+from transformers.tokenization_utils import PaddingStrategy, TruncationStrategy
+from transformers import AutoTokenizer, ChameleonProcessor
+from align_anything.models.chameleon_model import AccustomedChameleonModel
+
+from align_anything.utils.multi_process import get_current_device
+from align_anything.utils.template_registry import get_template_class
+from align_anything.utils.tools import right_padding
+from datasets import load_dataset
+
+ALLOWED_ATTRIBUTES = ['split_token']
+DEFAULT_SPLIT_TOKEN = 'ASSISTANT:'
+IGNORE_INDEX = -100
+
+
+def load_image(image_path: str):
+    try:
+        if image_path.startswith("http"):
+            image = Image.open(requests.get(image_path, stream=True).raw)
+        else:
+            image = Image.open(image_path)
+        return image
+    except Exception as e:
+        print(f"Error occured when dealing with {image_path}")
+        raise Exception
+
+def insert_img_token(text, image):
+    # do the same for worse
+    if isinstance(image, str):
+        decoded_images = [load_image(image)]
+        num_images = 1
+    elif isinstance(image, list):
+        decoded_images = [load_image(img) for img in image]
+        num_images = len(image)
+    elif isinstance(image, Image.Image):
+        decoded_images = [image]
+        num_images = 1
+    else:
+        num_images = 0
+        decoded_images = None
+    
+    processed_text = f"{'<image>' * num_images}{text}"
+    return processed_text, decoded_images
+
+def safe_add(list1, list2):
+    if list1 is None and list2 is None:
+        return []
+    elif list1 is None:
+        return list2.copy()
+    elif list2 is None:
+        return list1.copy()
+    else:
+        return list1 + list2
+    
+def format_sample_cham(raw_sample: dict[str, Any]) -> dict[str, Any]:
```

Comment:
Fixed in https://github.com/PKU-Alignment/align-anything/pull/46/commits/71067a935888c0d75b5003674a395464a0b2cfb2 by adding a docstring to every format_sample function. Hope this would make the code easier to read and clarify some misunderstandings.